File: demo-app/src/demo_app/defects.py
"""Intentional-defect flag registry.

Reads CSV ids from ``DEFECTS`` env var and branches handlers to introduce
one bug per id. Resolved per-call so tests can mutate ``os.environ``
between cases without re-importing.

Catalog: ``docs/defects/``.
"""

from __future__ import annotations

import os
from typing import Final

KNOWN_DEFECTS: Final[frozenset[str]] = frozenset(
    {
        "login_accepts_any_pin",
        "negative_qty_allowed",
        "off_by_one_pagination",
        "delete_skips_auth",
        "slow_query",
    }
)


def _active_set() -> frozenset[str]:
    raw = os.environ.get("DEFECTS", "").strip()
    if not raw:
        return frozenset()
    return frozenset(part.strip() for part in raw.split(",") if part.strip())


def enabled(defect_id: str) -> bool:
    """True iff ``defect_id`` is in ``KNOWN_DEFECTS`` and currently active.

    Unknown ids always return False, so a typo can't silently flip a
    different defect on.
    """

    if defect_id not in KNOWN_DEFECTS:
        return False
    return defect_id in _active_set()


def active() -> frozenset[str]:
    return _active_set() & KNOWN_DEFECTS

File: web/src/lib/defects.ts
// Intentional-defect flag registry for the in-browser SUT.
// Mirrors `demo-app/src/demo_app/defects.py`.
//
// Active ids come from sessionStorage (DefectsPanel toggles) merged with
// the `VITE_DEFECTS` build-time default. MSW handlers must call
// `defectEnabled()` *inside* the handler body, not at import time, so
// toggles take effect on the next request without a reload.

export const KNOWN_DEFECTS = [
  "login_accepts_any_pin",
  "negative_qty_allowed",
  "off_by_one_pagination",
  "delete_skips_auth",
  "slow_query",
] as const;

export type DefectId = (typeof KNOWN_DEFECTS)[number];

const STORAGE_KEY = "qa-automation-lab.defects";
const KNOWN_SET = new Set<string>(KNOWN_DEFECTS);

function parseCsv(raw: string | null | undefined): Set<string> {
  if (!raw) return new Set();
  return new Set(
    raw
      .split(",")
      .map((s) => s.trim())
      .filter(Boolean),
  );
}

function sessionDefects(): Set<string> {
  if (typeof window === "undefined") return new Set();
  try {
    return parseCsv(window.sessionStorage.getItem(STORAGE_KEY));
  } catch {
    return new Set();
  }
}

function buildDefects(): Set<string> {
  return parseCsv(import.meta.env?.VITE_DEFECTS);
}

export function activeDefects(): DefectId[] {
  const seen = new Set<string>([...buildDefects(), ...sessionDefects()]);
  return KNOWN_DEFECTS.filter((id) => seen.has(id));
}

export function defectEnabled(id: DefectId): boolean {
  if (!KNOWN_SET.has(id)) return false;
  return activeDefects().includes(id);
}

export function setDefects(ids: readonly DefectId[]): void {
  if (typeof window === "undefined") return;
  const csv = ids.filter((id) => KNOWN_SET.has(id)).join(",");
  try {
    if (csv) {
      window.sessionStorage.setItem(STORAGE_KEY, csv);
    } else {
      window.sessionStorage.removeItem(STORAGE_KEY);
    }
  } catch {
    /* sessionStorage disabled — visitor can still toggle in this tab */
  }
}
